### PR TITLE
ci: Versioned docs via mike

### PR DESCRIPTION
**Pull request by jkayani**
### Description

Version docs via `mike`. Creates 2 moving aliases: `latest`, for the main branch of docs, and `stable` for the latest tagged release. Versions are published and aliases are updated on tag push/main push. Aliases are used off the GH pages URL to redirect to the version they point to.

Live example: https://jkayani.github.io/argocd-vault-plugin/

**Fixes:** #194 

### Checklist
Please make sure that your PR fulfills the following requirements:
- [ ] Reviewed the guidelines for contributing to this repository
- [ ] The commit message follows the [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] Tests for the changes have been updated
- [ ] Docs have been added / updated

### Type of Change

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [x] Build/CI related changes
- [x] Documentation content changes
- [ ] Other (please describe)

### CODE REVIEW

File: .github/workflows/mkdocs.yml [hunk 1]
```diff
@@ -11,5 +13,10 @@ jobs:
       - uses: actions/setup-python@v2
         with:
           python-version: 3.x
-      - run: pip install mkdocs-material 
-      - run: mkdocs gh-deploy --force
+      - run: |
+          pip install mkdocs-material 
+          pip install mike
+      - run: | 
+        export REF=$(echo $GITHUB_REF | pcregrep --only-matching=1 ".*\/.*\/(.*)")
```
Comment (jkayani) [inline .github/workflows/mkdocs.yml:20]: This shell snippet works great locally (setting `GITHUB_REF` to stuff like `refs/heads/main` and `refs/tags/v1.2.3`). Need to be sure if this works in Actions though

Comment (jkayani) [reply .github/workflows/mkdocs.yml]: Tested here: https://github.com/jkayani/argocd-vault-plugin/runs/3561014531?check_suite_focus=true

**Comment by jkayani** [comment]
The only complication is how to handle merging this, since we want docs based off v1.3.1 as the default (`stable`). But those can't be generated via CI since v1.3.1 was released already.

So we just have to do it manually:

- Apply the patch to `mkdocs.yaml` so that the version dropdown appears when generating the docs

- Run:
```
mike delete --all # to delete the existing HTML in gh-pages branch since we don't need it anymore
git checkout v1.3.1
mike deploy -u v1.3.1 stable

mike set-default stable
git push --force origin gh-pages:gh-pages
```

so that we add a v1.3.1 (alias `stable`) which is the default version of the docs to `gh-pages` branch (example commit: https://github.com/jkayani/argocd-vault-plugin/commit/ee051fad4c3e5d2efeb93de4844ecfd246c1880a)

- Then, merge this PR. It will run the Mkdocs action which will push a commit to `gh-pages` adding a `main` version (alias `latest`) on top. (example CI run: https://github.com/jkayani/argocd-vault-plugin/actions/runs/1219046108)

Result is versioned docs as if we had them the whole time: https://jkayani.github.io/argocd-vault-plugin. The GH pages URL goes to the default version; appending an alias redirects you to whatever version the alias points to

**Comment by jkayani** [comment]
Proof it works on tags:

CI run: https://github.com/jkayani/argocd-vault-plugin/runs/3561655188?check_suite_focus=true

End result: https://jkayani.github.io/argocd-vault-plugin now points to v1.4.0 docs

**Comment by werne2j** [comment]
@jkayani we will have to manually push each of the tags as well as 1.3.1, right?

**Review by werne2j — APPROVED** [review]
Looks good!

**Comment by jkayani** [comment]
> @jkayani we will have to manually push each of the tags as well as 1.3.1, right?

If we wanted to retroactively release versioned docs, yes. But I don't know if that's really worthwhile - I think having the current release people are intended to use + the "preview" docs are better

> So why even do latest. Let’s just build the docs on a tag

I think it's worth having so a.) we can proof-read docs/make sure Mkdocs formats stuff right and b.) point people who need unreleased but available feature X to the docs

<hr/>

With all the above I think it might better to change things slightly: 

- alias the latest tagged release as `latest` (or just omit the alias, since we can just set the latest tagged version vx.y.yz as the default)
- alias the `main` based docs as `preview` 

That way it's more clear which is which. Thoughts?

**Comment by werne2j** [comment]
> If we wanted to retroactively release docs for previous versions, yes. But I don't know if that's really worthwhile - I think having the current release people are intended to use + the "preview" docs is enough

I would like to be it retroactively, at least for the tags that have the docs.

I am fine with the other suggestions

**Comment by jkayani** [comment]
Going to try to add a "Your looking at unreleased docs" on the preview version, and write a script to generate versioned docs for our previous releases (where applicable)

**Review by werne2j — APPROVED** [review]

**Comment by jkayani** [comment]
- Updated PR to include a banner saying the `main` docs are for unreleased features and a link to the GH Pages URL which is set to be default, example here: https://jkayani.github.io/argocd-vault-plugin/main/
- Kept the `stable` alias so that even when we release new tags, the `stable` alias points to the latest tag and we only have to run `mike set-default` once when setting things up

**Comment by jkayani** [comment]
Before merging this PR I will run:

```
mike delete --all
for tag that has mkdocs:
  git checkout $tag
  git cherry-pick versioned-docs
  mike deploy -u $tag stable
# manually inspect to make sure generated HTML looks right
mike set-default stable
git push --force origin gh-pages:gh-pages
```

And then this PR will create our `preview` docs and automate our Mkdocs workflow to do the above on tagged releases/merges to main

**Comment by jkayani** [comment]
Versioned docs are pushed: https://github.com/IBM/argocd-vault-plugin/commits/gh-pages. This PR is ready to merge I think

**Review by werne2j — APPROVED** [review]

**Pull request merged**